### PR TITLE
[AINode] Fix bug of sundial and forecast udf

### PR DESCRIPTION
**Pull request by yunbow30944**
## Description

+ Fix bug that forecast udf cannot be called
+ Fix bug of sundial model

### Before

屏幕截图 2025-11-15 222816

### After 
屏幕截图 2025-11-17 143001

### CODE REVIEW

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py [hunk 1]
```diff
@@ -603,35 +603,42 @@ def prepare_inputs_for_generation(
...
+            if attention_mask is not None and attention_mask.shape[1] > token_num:
                 input_ids = input_ids[:, -(attention_mask.shape[1] - past_length) :]
             # 2 - If the past_length is smaller than input_ids', then input_ids holds all input tokens. We can discard
             # input_ids based on the past_length.
-            elif past_length < (input_ids.shape[1] // self.config.input_token_len):
-                input_ids = input_ids[:, past_length * self.config.input_token_len :]
+            elif past_length < token_num:
+                # TODO: Actually, we need to know the output_token_lens used in the last generation step.
+                #  Sundial will pad the input when it is non-divisible, so we cannot use past_length to slice input_ids
+                input_ids = input_ids[:, -self.config.output_token_lens[0] :]
```
Comment (CRZbulabula) [inline iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py]: Why change the computation of `input_ids` here?

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py [hunk 2]
```diff
@@ -603,35 +603,42 @@ def prepare_inputs_for_generation(
...
             position_ids = attention_mask.long().cumsum(-1) - 1
             position_ids.masked_fill_(attention_mask == 0, 1)
-            if past_key_values:
-                position_ids = position_ids[
-                    :, -(input_ids.shape[1] // self.config.input_token_len) :
-                ]
+            if past_key_values is not None and past_length > 0:
+                token_num = (
+                    input_ids.shape[1] + self.config.input_token_len - 1
+                ) // self.config.input_token_len
```
Comment (CRZbulabula) [inline iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py:638]: Why re-calculate the `token_num` again? Plz make an explanation.

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py (hunk 1 above)
Comment (RkGrit) [reply iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py]: +1.
Why not use " past_length * self.config.input_token_len " ？
"input_ids" should not be truncated by "output_token_lens" because the lengths of the input and output are not correlated.

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py (hunk 2 above)
Comment (yunbow30944) [reply iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py:638]: Because the input_ids is sliced, so the position_ids should be consistent with the new length of input.

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py (hunk 1 above)
Comment (yunbow30944) [reply iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py]: Because Sundial will pad the input when it is non-divisible, so we cannot use past_length to slice input_ids, it will slice more sequence than we expected.

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py [hunk 3]
```diff
@@ -616,7 +616,11 @@ def prepare_inputs_for_generation(
             if attention_mask is not None and attention_mask.shape[1] > (
                 input_ids.shape[1] // self.config.input_token_len
```
Comment (RkGrit) [inline iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py:617]: When calculating the token length of input_ids, since Sundial performs padding in the embedding stage, fragment cannot be discarded but should be rounded up. Like：
token_num = (
                    input_ids.shape[1] + self.config.input_token_len - 1
                ) // self.config.input_token_len

File: iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py [hunk 4]
```diff
@@ -616,7 +616,11 @@ def prepare_inputs_for_generation(
...
             ):
-                input_ids = input_ids[:, -(attention_mask.shape[1] - past_length) :]
+                input_ids = input_ids[
+                    :,
+                    -(attention_mask.shape[1] - past_length)
+                    * self.config.input_token_len :,
+                ]
             # 2 - If the past_length is smaller than input_ids', then input_ids holds all input tokens. We can discard
             # input_ids based on the past_length.
             elif past_length < (input_ids.shape[1] // self.config.input_token_len):
```
Comment (RkGrit) [inline iotdb-core/ainode/iotdb/ainode/core/model/sundial/modeling_sundial.py:626]: The same as above

File: iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py [hunk 5]
```diff
@@ -606,7 +606,11 @@ def prepare_inputs_for_generation(
             if attention_mask is not None and attention_mask.shape[1] > (
                 input_ids.shape[1] // self.config.input_token_len
```
Comment (RkGrit) [inline iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py:607]: The same as above

File: iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py [hunk 6]
```diff
@@ -606,7 +606,11 @@ def prepare_inputs_for_generation(
...
             ):
-                input_ids = input_ids[:, -(attention_mask.shape[1] - past_length) :]
+                input_ids = input_ids[
+                    :,
+                    -(attention_mask.shape[1] - past_length)
+                    * self.config.input_token_len :,
+                ]
             # 2 - If the past_length is smaller than input_ids', then input_ids holds all input tokens. We can discard
             # input_ids based on the past_length.
             elif past_length < (input_ids.shape[1] // self.config.input_token_len):
```
Comment (RkGrit) [inline iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py:616]: The same as above

**Review by RkGrit — COMMENTED** [review]
PTAL

File: iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py (hunk 6 above)
Comment (RkGrit) [reply iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py:616]: The same as above

image

Comment (yunbow30944) [reply iotdb-core/ainode/iotdb/ainode/core/model/timerxl/modeling_timer.py:616]: timerxl will truncate the input at the begining of inference, so it doesnt matter
